mono/profiler/mono-profiler-malloc.c

```diff
@@ -0,0 +1,1354 @@
+	alloc_unlock ();
+}
+
+static void
+memdom_alloc (MonoProfiler *prof, void* memdom, size_t size, const char *tag)
+{
+	alloc_lock ();
+	MemDomInfo *info = hashtable_find (&memdom_table, memdom);
+	if (info)
+		update_tag (&info->tags, tag, size, 0);
+	alloc_unlock ();
+	if (PRINT_RT_ALLOC)
+		printf ("memdom %p alloc %zu %s\n", memdom, size, tag);
+}
+
+static const char*
+filter_bt (int skip, const char *filter_funcs)
+{
+	void *bt_ptrs [10];
+ 	int c = backtrace (bt_ptrs, 10);
+	if (c) {
+		int i;
+		const char *it = NULL;
+		for (i = 0; i < c - (skip + 1); ++i) {
+			Dl_info info;
+			if (!dladdr (bt_ptrs [skip + i], &info))
+				continue;
+			if (strstr (info.dli_sname, "_malloc") || strstr (info.dli_sname, "report_alloc"))
+				continue;
+			if (strstr (info.dli_sname, filter_funcs))
+				continue;
+			it = info.dli_sname;
+			break;
+		}
+		return it;
+	}
+	return NULL;
+}
+
+static void
+runtime_malloc_event (MonoProfiler *prof, void *address, size_t size, const char *tag)
+{
+	alloc_lock ();
+	AllocInfo *info = hashtable_find (&alloc_table, address);
+	if (!info) {
+		printf ("stray alloc that didn't come from g_malloc %p\n", address);
+		break_on_bad_runtime_malloc ();
+		goto done;
+	}
+
+	if (info->tag) {
+		printf ("runtime reported same pointer twice %p %s x %s\n", address, info->tag, tag);
+		break_on_bad_runtime_malloc ();
+		goto done;
+	}
+
+	if (info->size != size) {
+		printf ("runtime reported pointer with different sizes %p %zu x %zu\n", address, info->size, size);
+		break_on_bad_runtime_malloc ();
+		goto done;
+	}
+
+	info->tag = tag;
+	update_tag (&malloc_tags, tag, info->size, info->waste);
+
+
+	if (info->alloc_func) {
+		//we got it reported, take away from the stray list
+		update_tag (&malloc_tags_by_fun, info->alloc_func, -info->size, -info->waste);
+		info->alloc_func = NULL;
+	}
+
+done:
+	alloc_unlock ();
+}
+
+enum {
+	RECORD_VALLOC = 1 << 10,
+	RECORD_VFREE = 1 << 11,
+	RECORD_MPROTECT = 1 << 12,
+};
+
+typedef struct {
+	char *addr_start;
+	char *addr_end;
+	const char *tag;
+	int flags;
+} VAllocRecord;
+
+#define VALLOC_ENTRIES (4096 / sizeof (VAllocRecord) - 1)
+
+typedef struct _VAllocRecordBucket VAllocRecordBucket;
+
+struct _VAllocRecordBucket {
+	int index;
+	VAllocRecordBucket *next_bucket;
+	VAllocRecord entries [VALLOC_ENTRIES];
+};
+
+static volatile int valloc_log_state;
+
+static void __attribute__((noinline))
+break_on_fail (void)
+{
+}
+
+
+#define FAIL(ERR) do { printf ("FAILED: %s\n", ERR); break_on_fail (); exit (-1); } while (0)
+
+static gboolean
+lock_exclusive (void)
+{
+	do {
+		/*
+		A positive value means the lock is taken.
+		TODO: Have a write-requested state to avoid writer starvation
+		*/
+		if (valloc_log_state == -1)
+			FAIL ("Somebody messed with the exclusive lock, already locked");
+
+		if (valloc_log_state) {
+			usleep (1);
+			continue;
+		}
+	} while (__sync_val_compare_and_swap (&valloc_log_state, 0, -1) != 0);
+	__sync_synchronize ();
```

Comment:
You can use `mono/utils/atomic.h` and `mono/utils/mono-membar.h` from profiler modules.